File: statefun-examples/statefun-ridesharing-example/statefun-ridesharing-example-simulator/src/main/java/org/apache/flink/statefun/examples/ridesharing/simulator/simulation/Passenger.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.apache.flink.statefun.examples.ridesharing.simulator.simulation;

import lombok.extern.slf4j.Slf4j;
import org.apache.flink.statefun.examples.ridesharing.generated.InboundPassengerMessage;
import org.apache.flink.statefun.examples.ridesharing.generated.OutboundPassengerMessage;
import org.apache.flink.statefun.examples.ridesharing.simulator.model.WebsocketPassengerEvent;
import org.apache.flink.statefun.examples.ridesharing.simulator.simulation.engine.LifecycleMessages;
import org.apache.flink.statefun.examples.ridesharing.simulator.simulation.engine.Simulatee;

@Slf4j
public class Passenger implements Simulatee {
  private final PassengerMessaging simulation;
  private final StateMachine<WebsocketPassengerEvent.PassengerStatus> stateMachine;
  private final String id;
  private final int startCell;
  private final int endCell;

  private String driverId;

  Passenger(PassengerMessaging simulation, String id, int startCell, int endCell) {
    this.simulation = simulation;
    this.id = id;
    this.startCell = startCell;
    this.endCell = endCell;
    this.stateMachine = passengerStateMachine();
  }

  /**
   * when this state machine initializes we send a ride request, and move to the next state ({@link
   * WebsocketPassengerEvent.PassengerStatus#REQUESTING}
   */
  @SuppressWarnings("unused")
  private WebsocketPassengerEvent.PassengerStatus init(LifecycleMessages.Initialization ignored) {

    final InboundPassengerMessage rideRequest =
        InboundPassengerMessage.newBuilder()
            .setPassengerId(id)
            .setRequestRide(
                InboundPassengerMessage.RequestRide.newBuilder()
                    .setStartGeoCell(startCell)
                    .setEndGeoCell(endCell))
            .build();

    // send to application
    simulation.outgoingPassengerEvent(rideRequest);

    // send to the web socket
    simulation.broadcastPassengerSimulationEvent(
        WebsocketPassengerEvent.builder()
            .passengerId(id)
            .startLocation(startCell)
            .endLocation(endCell)
            .status(WebsocketPassengerEvent.PassengerStatus.REQUESTING)
            .rideId("") // we don't have it yet.
            .build());

    // next state would be requesting a ride
    return WebsocketPassengerEvent.PassengerStatus.REQUESTING;
  }

  /** The ride failed, couldn't find a driver nearby. */
  private WebsocketPassengerEvent.PassengerStatus rideFailed(OutboundPassengerMessage message) {
    final String failedRideId = message.getRideFailed().getRideId();
    //
    // notify the websocket
    //
    simulation.broadcastPassengerSimulationEvent(
        WebsocketPassengerEvent.builder()
            .passengerId(id)
            .startLocation(startCell)
            .endLocation(endCell)
            .status(WebsocketPassengerEvent.PassengerStatus.FAIL)
            .rideId(failedRideId)
            .build());

    return WebsocketPassengerEvent.PassengerStatus.FAIL;
  }

  /** A driver was found, now waiting for the pickup to happen. */
  private WebsocketPassengerEvent.PassengerStatus driverFound(OutboundPassengerMessage message) {
    OutboundPassengerMessage.DriverHasBeenFound driverFound = message.getDriverFound();

    simulation.broadcastPassengerSimulationEvent(
        WebsocketPassengerEvent.builder()
            .rideId("") // TODO: ?
            .passengerId(id)
            .startLocation(startCell)
            .endLocation(endCell)
            .driverId(driverFound.getDriverId())
            .status(WebsocketPassengerEvent.PassengerStatus.WAITING_FOR_RIDE_TO_START)
            .build());

    // capture the driver id
    driverId = driverFound.getDriverId();

    return WebsocketPassengerEvent.PassengerStatus.WAITING_FOR_RIDE_TO_START;
  }

  private WebsocketPassengerEvent.PassengerStatus rideStarted(OutboundPassengerMessage message) {
    OutboundPassengerMessage.RideStarted rideStarted = message.getRideStarted();

    simulation.broadcastPassengerSimulationEvent(
        WebsocketPassengerEvent.builder()
            .rideId("") // TODO: ?
            .passengerId(id)
            .startLocation(startCell)
            .endLocation(endCell)
            .driverId(rideStarted.getDriverId())
            .status(WebsocketPassengerEvent.PassengerStatus.RIDING)
            .build());

    return WebsocketPassengerEvent.PassengerStatus.RIDING;
  }

  @SuppressWarnings("unused")
  private WebsocketPassengerEvent.PassengerStatus rideEnded(OutboundPassengerMessage ignored) {
    simulation.broadcastPassengerSimulationEvent(
        WebsocketPassengerEvent.builder()
            .rideId("") // TODO: ?
            .passengerId(id)
            .startLocation(startCell)
            .endLocation(endCell)
            .driverId(driverId)
            .status(WebsocketPassengerEvent.PassengerStatus.DONE)
            .build());

    return WebsocketPassengerEvent.PassengerStatus.DONE;
  }

  private StateMachine<WebsocketPassengerEvent.PassengerStatus> passengerStateMachine() {
    StateMachine<WebsocketPassengerEvent.PassengerStatus> stateMachine =
        new StateMachine<>(WebsocketPassengerEvent.PassengerStatus.IDLE);

    stateMachine.withTerminalState(WebsocketPassengerEvent.PassengerStatus.FAIL);
    stateMachine.withTerminalState(WebsocketPassengerEvent.PassengerStatus.DONE);

    // send the req
    stateMachine.withState(
        WebsocketPassengerEvent.PassengerStatus.IDLE,
        LifecycleMessages.Initialization.class,
        this::init);

    // req success
    stateMachine.withState(
        WebsocketPassengerEvent.PassengerStatus.REQUESTING,
        OutboundPassengerMessage.class,
        OutboundPassengerMessage::hasDriverFound,
        this::driverFound);

    // req failure (terminal)
    stateMachine.withState(
        WebsocketPassengerEvent.PassengerStatus.REQUESTING,
        OutboundPassengerMessage.class,
        OutboundPassengerMessage::hasRideFailed,
        this::rideFailed);

    // ride started
    stateMachine.withState(
        WebsocketPassengerEvent.PassengerStatus.WAITING_FOR_RIDE_TO_START,
        OutboundPassengerMessage.class,
        OutboundPassengerMessage::hasRideStarted,
        this::rideStarted);

    // ride finished (terminal)
    stateMachine.withState(
        WebsocketPassengerEvent.PassengerStatus.RIDING,
        OutboundPassengerMessage.class,
        OutboundPassengerMessage::hasRideEnded,
        this::rideEnded);

    return stateMachine;
  }

  @Override
  public String id() {
    return id;
  }

  @Override
  public boolean isDone() {
    return stateMachine.isAtTerminalState();
  }

  @Override
  public void apply(Object event) {
    stateMachine.apply(event);
  }

  @Override
  public boolean needReschedule() {
    return false;
  }
}

File: statefun-go-sdk/examples/rideshare/passenger.go
// Licensed to the Apache Software Foundation (ASF) under one or more
// contributor license agreements.  See the NOTICE file distributed with
// this work for additional information regarding copyright ownership.
// The ASF licenses this file to You under the Apache License, Version 2.0
// (the "License"); you may not use this file except in compliance with
// the License.  You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package main

import (
	"context"
	"fmt"
	"github.com/apache/flink-statefun/statefun-go-sdk/v2/pkg/flink/statefun"
	"github.com/apache/flink-statefun/statefun-go-sdk/v2/pkg/flink/statefun/io"
	"google.golang.org/protobuf/types/known/anypb"
	"math/rand"
)

type Passenger struct{}

func (p *Passenger) Invoke(ctx context.Context, out statefun.Output, msg *anypb.Any) error {
	passengerId := statefun.Self(ctx).Id

	var inbound InboundPassengerMessage
	if err := msg.UnmarshalTo(&inbound); err == nil {
		request := inbound.GetRequestRide()
		return p.whenRideRequested(passengerId, out, request)
	}

	var driverJoin DriverJoinsRide
	if err := msg.UnmarshalTo(&driverJoin); err == nil {
		return p.whenDriverJoins(passengerId, out, &driverJoin)
	}

	var rideFailed RideFailed
	if err := msg.UnmarshalTo(&rideFailed); err == nil {
		return p.whenRideFailed(passengerId, out, &rideFailed)
	}

	var rideStarted RideStarted
	if err := msg.UnmarshalTo(&rideStarted); err == nil {
		return p.whenRideHasStarted(passengerId, out, &rideStarted)
	}

	var rideEnded RideEnded
	if err := msg.UnmarshalTo(&rideEnded); err == nil {
		return p.whenRideHasEnded(passengerId, out)
	}

	return nil
}

func (p *Passenger) whenRideRequested(passengerId string, out statefun.Output, request *InboundPassengerMessage_RequestRide) error {
	rideId := fmt.Sprintf("ride-%d", rand.Uint64())

	joinRide := &PassengerJoinsRide{
		PassengerId:  passengerId,
		StartGeoCell: request.GetStartGeoCell(),
		EndGeoCell:   request.GetEndGeoCell(),
	}

	ride := statefun.Address{
		FunctionType: RideType,
		Id:           rideId,
	}

	return out.Send(ride, joinRide)
}

func (p *Passenger) whenDriverJoins(passengerId string, out statefun.Output, driverJoin *DriverJoinsRide) error {
	record := io.KafkaRecord{
		Topic: "to-rPassenger",
		Key:   passengerId,
		Value: &OutboundPassengerMessage{
			PassengerId: passengerId,
			Message: &OutboundPassengerMessage_DriverFound{
				DriverFound: &OutboundPassengerMessage_DriverHasBeenFound{
					DriverId:      driverJoin.DriverId,
					DriverGeoCell: driverJoin.DriverLocation,
				},
			},
		},
	}

	if message, err := record.ToMessage(); err != nil {
		return err
	} else {
		return out.SendEgress(ToPassengerEgress, message)
	}
}

func (p *Passenger) whenRideFailed(passengerId string, out statefun.Output, rideFailed *RideFailed) error {
	record := io.KafkaRecord{
		Topic: "to-passenger",
		Key:   passengerId,
		Value: &OutboundPassengerMessage{
			PassengerId: passengerId,
			Message: &OutboundPassengerMessage_RideFailed_{
				RideFailed: &OutboundPassengerMessage_RideFailed{
					RideId: rideFailed.RideId,
				},
			},
		},
	}

	if message, err := record.ToMessage(); err != nil {
		return err
	} else {
		return out.SendEgress(ToPassengerEgress, message)
	}
}

func (p *Passenger) whenRideHasStarted(passengerId string, out statefun.Output, started *RideStarted) error {
	record := io.KafkaRecord{
		Topic: "to-passenger",
		Key:   passengerId,
		Value: &OutboundPassengerMessage{
			PassengerId: passengerId,
			Message: &OutboundPassengerMessage_RideStarted_{
				RideStarted: &OutboundPassengerMessage_RideStarted{
					DriverId: started.DriverId,
				},
			},
		},
	}

	if message, err := record.ToMessage(); err != nil {
		return err
	} else {
		return out.SendEgress(ToPassengerEgress, message)
	}
}

func (p *Passenger) whenRideHasEnded(passengerId string, out statefun.Output) error {
	record := io.KafkaRecord{
		Topic: "to-passenger",
		Key:   passengerId,
		Value: &OutboundPassengerMessage{
			PassengerId: passengerId,
			Message: &OutboundPassengerMessage_RideEnded_{
				RideEnded: &OutboundPassengerMessage_RideEnded{},
			},
		},
	}

	if message, err := record.ToMessage(); err != nil {
		return err
	} else {
		return out.SendEgress(ToPassengerEgress, message)
	}
}
